### PR TITLE
Trigger a reroute after registering a node for removal

### PR DESCRIPTION
**Pull request by AthenaEryma**
The cluster will not automatically react to the node shutdown being
registered unless we notify it that something has changed that may
require a change in shard allocation. This commit modifies the Put
Shutdown action to invoke a reroute after the cluster state has been
updated, as well as an integration test to verify that shards quickly
move away from nodes which are shutting down for removal.

### CODE REVIEW

**Comment by dakrone** [comment]
@gwbrown I think we also need one for removing a node from the shutdown list? We want to make sure shards can go back to a node if it's done shutting down.

**Comment by gwbrown** [comment]
@dakrone Good call, added reroute call after deletion and another test to make sure it works.

File: x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java [hunk 1]
```diff
@@ -77,6 +82,22 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                         .build();
 
                 }
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
```
Comment (dakrone) [inline x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java:87]: Do we have to do this after the cluster state has been processed? I think I would prefer if possible to do it in a single cluster state update, like with https://github.com/elastic/elasticsearch/blob/eee399aeb5825fc3a98ec62f719ee41d584b4422/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java#L238

Is there a reason we shouldn't inject the `AllocationService` into these transport actions and then if a node was added/removed we can call the reroute from within the `execute` method and only generate a single cluster state update.

What do you think?

**Review by dakrone — CHANGES_REQUESTED** [review]
I left a comment about potentially doing this in a single cluster state update. Let me know what you think!

File: x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java:87]: The advantage of using the reroute service here is that it batches reroutes together. Reroutes are expensive (and idempotent) so batching is preferable even if it costs an extra cluster state update sometimes. If you put/delete a bunch of these shutdown markers all at once then we don't want to do a reroute for each one, and if there are shards starting/failing at the same time then we can piggy-back our reroute on theirs.

I think it'd be better not to notify the listener until after the reroute completes, however: that way we should be able to assert that the shards already started moving when the response comes in.

Comment (dakrone) [reply x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java:87]: Okay, I didn't realize that reroutes were expensive and that they could be batched, in that case, doing the separate update is better as David said. +1 on waiting until the reroute completes before returning the listener.

Comment (gwbrown) [reply x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java]: I originally went this route based on advice that using the reroute service here was better than allocation service, though I didn't know the reasoning why. Thanks Dave!

Updated this to wait until the reroute has been successfully kicked off to respond to the request (after spending way too long trying to figure out why the request kept timing out in the tests, only to realize I forgot the `listener.onResponse(AcknowledgedResponse.TRUE);` in the PUT action, oops).

Comment (DaveCTurner) [reply x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java]: ++ I opened https://github.com/elastic/elasticsearch/pull/73512 to record the information about rerouting somewhere easier for others to find.

**Review by dakrone — APPROVED** [review]
LGTM, thanks for iterating on this!

**Pull request merged**